File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
#         s = ""
#         while head :
#             s += str(head.val)
#             head = head.next
        
#         return s == s[::-1]

        prev = None
        slow = fast = head
        
        while fast and fast.next :
            fast = fast.next.next
            temp = slow.next
            slow.next = prev
            prev = slow
            slow = temp
        
        if fast :
            slow = slow.next
        
        while prev :
            if slow.val != prev.val :
                return False
            prev , slow = prev.next , slow.next
        
        return True

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

import java.util.*;
class Solution {
    public boolean isPalindrome(ListNode head) {
        String s = "";
        while(head != null){
            s += (char)(head.val + '0');
            head = head.next;
        }
        int i = 0 , j = s.length()-1;
        while(i <= j){
            if(s.charAt(i++)!=s.charAt(j--)){
                return false;
            }
        }
        return true;
    }
}